src/daft-plan/src/optimization/rules/push_down_projection.rs

```diff
@@ -0,0 +1,398 @@
+                if maybe_new_left_upstream
+                    .as_ref()
+                    .or(maybe_new_right_upstream.as_ref())
+                    .is_some()
+                {
+                    let new_left_upstream = maybe_new_left_upstream.unwrap_or(join.input.clone());
+                    let new_right_upstream = maybe_new_right_upstream.unwrap_or(join.right.clone());
+                    let new_join =
+                        upstream_plan.with_new_children(&[new_left_upstream, new_right_upstream]);
+                    let new_plan = plan.with_new_children(&[new_join]);
+                    // Retry optimization now that the upstream node is different.
+                    let new_plan = self.try_optimize(&new_plan)?.unwrap_or(new_plan);
+                    Ok(Some(new_plan))
+                } else {
+                    Ok(None)
+                }
+            }
+            LogicalPlan::Distinct(_) => {
+                // Cannot push down past a Distinct,
+                // since Distinct implicitly requires all parent columns.
+                Ok(None)
+            }
+            LogicalPlan::Sink(_) => {
+                panic!("Bad projection due to upstream sink node: {:?}", projection)
+            }
+        }
+    }
+
+    fn try_optimize_aggregation(
+        &self,
+        aggregation: &Aggregate,
+        plan: &LogicalPlan,
+    ) -> DaftResult<Option<Arc<LogicalPlan>>> {
+        // If this aggregation prunes columns from its upstream,
+        // then explicitly create a projection to do so.
+        let upstream_plan = &aggregation.input;
+        let upstream_schema = upstream_plan.schema();
+
+        let aggregation_required_cols = &plan.required_columns()[0];
+        if aggregation_required_cols.len() < upstream_schema.names().len() {
+            let new_subprojection: LogicalPlan = {
+                let pushdown_column_exprs = aggregation_required_cols
+                    .iter()
+                    .map(|s| Expr::Column(s.clone().into()))
+                    .collect::<Vec<_>>();
+
+                Project::try_new(
+                    upstream_plan.clone(),
+                    pushdown_column_exprs,
+                    Default::default(),
+                )?
+                .into()
+            };
+
+            let new_aggregation = plan.with_new_children(&[new_subprojection.into()]);
+            Ok(Some(new_aggregation))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+impl OptimizerRule for PushDownProjection {
+    fn apply_order(&self) -> Option<ApplyOrder> {
+        Some(ApplyOrder::TopDown)
+    }
+
+    fn try_optimize(&self, plan: &LogicalPlan) -> DaftResult<Option<Arc<LogicalPlan>>> {
+        match plan {
+            LogicalPlan::Project(projection) => self.try_optimize_project(projection, plan),
+            // Aggregations also do column projection
+            LogicalPlan::Aggregate(aggregation) => self.try_optimize_aggregation(aggregation, plan),
```

Comment:
Nice, I like this splitting into two methods for each of the root node types under consideration!